calendar_af/models/calendar_appointment.py

```diff
@@ -0,0 +1,636 @@
+        # remove one day from start date since we add a day at the start of each loop.
+        loop_start = datetime.now() - timedelta(days=1)
+        i = 0
+
+        while i < days_first:
+            loop_start = loop_start + timedelta(days=1)
+            if (
+                loop_start.weekday() in [0, 1, 2, 3, 4]
+            ) and self._check_resource_calendar_date(loop_start):
+                i += 1
+
+        return loop_start.replace(
+            hour=BASE_DAY_START.hour,
+            minute=BASE_DAY_START.minute,
+            second=0,
+            microsecond=0,
+        )
+
+    def stop_meeting_search(self, start_meeting_search, type_id):
+        days_last = self.type_id.days_last if self.type_id.days_last else 15
+        # remove one day from start date since we add a day at the start of each loop.
+        loop_start = start_meeting_search - timedelta(days=1)
+        i = 0
+
+        while i < days_last:
+            loop_start = loop_start + timedelta(days=1)
+            if (
+                loop_start.weekday() in [0, 1, 2, 3, 4]
+            ) and self._check_resource_calendar_date(loop_start):
+                i += 1
+
+        return loop_start.replace(
+            hour=BASE_DAY_STOP.hour,
+            minute=BASE_DAY_STOP.minute,
+            second=0,
+            microsecond=0,
+        )
+
+    @api.one
+    def inactivate(self, b=True):
+        """Inactivates self. Used as a workaround to inactivate from server actions."""
+        if b:
+            self.active = False
+        else:
+            self.active = True
+        return self.active
+
+    def generate_cancel_daily_note(self, cancel_reason, appointment):
+        pass
+
+    def cancel(self, cancel_reason):
+        """Cancels a planned meeting"""
+        # Perform access control.
+        if self.af_check_access():
+            # Checks passed. Run inner function with sudo.
+            return self.sudo()._cancel(cancel_reason)
+        raise Warning(_("You are not allowed to cancel these meetings."))
+
+    def _cancel(self, cancel_reason):
+        # Do not allow cancellation of meetings that have been sent to ACE
+        if not cancel_reason:
+            return False
+        for appointment in self:
+            if appointment.state == "confirmed":
+                appointment.state = "canceled"
+                appointment.cancel_reason = cancel_reason.id
+
+                self.generate_cancel_daily_note(cancel_reason, appointment)
+
+                appointment.occasion_ids = [(5, 0, 0)]
+
+                return True
+
+    def confirm_appointment(self):
+        """Confirm reserved booking"""
+        # Perform access control.
+        if self.af_check_access():
+            # Checks passed. Run inner function with sudo.
+            return self.sudo()._confirm_appointment()
+        raise Warning(_("You are not allowed to confirm these meetings."))
+
+    def _confirm_appointment(self):
+        for appointment in self:
```

Comment:
Is this for only one appointment or several appointments
If its only for one its OKish code.
If its for several appointment we only return the last ones status